tests/integration/test_adapter.py

```diff
@@ -0,0 +1,145 @@
+    """
+    try:
+        yield request.cls.tadapter(ParameterSet(), ParameterSet())
+    except TypeError:
+        pytest.skip("{} cannot be instantiated".format(str(request.cls.tadapter)))
+
+
+@pytest.fixture(scope="function")
+def test_config_paraset():
+    parameters = ParameterSet()
+    parameters.get_writable_scalar_view("ecs", ("World",), "K").set(3)
+    parameters.get_writable_scalar_view("rf2xco2", ("World",), "W / m^2").set(4.0)
+
+    yield parameters
+
+
+@pytest.fixture(scope="function")
+def test_drivers_core():
+    # doesn't exist yet but we'd want something like this for testing
+    core = rcp_26_core
+
+    yield core
+
+
+def test_adapter_base_class_init():
+    parametersstub = "Parameters"
+    outputstub = "Parameters"
+    Adapter.__abstractmethods__ = set()
+    adapter = Adapter(parametersstub, outputstub)  # pylint: disable=abstract-class-instantiated
+    assert adapter._parameters == parametersstub
+    assert adapter._output == outputstub
+
+
+def test_adapter_base_class_initialise_run_parameters():
+    Adapter.__abstractmethods__ = set()
+    adapter = Adapter(ParameterSet(), ParameterSet())  # noqa
+
+    start_time = 0
+    stop_time = 1
+    adapter.initialize_run_parameters(start_time, stop_time)
+    assert adapter._start_time == start_time
+    assert adapter._stop_time == stop_time
+
+
+def test_adapter_base_class_run():
+    Adapter.__abstractmethods__ = set()
+    adapter = Adapter(ParameterSet(), ParameterSet())  # noqa
+    start_time = 0
+    adapter.initialize_run_parameters(start_time, 1)
+    adapter.initialize_model_input()
+    adapter.reset()
+    assert adapter._current_time == start_time
+    adapter.run()
+    assert adapter.step() == start_time
+
+
+class _AdapterTester(object):
+    """Base class for adapter testing.
+
+    At minimum, a new adapter should define a subclass of this class called,
+    ``AdapterXTester`` which has ``tadapter`` equal to the adapter to be tested.
+    This ensures that the new adapter is subject to all of OpenSCM's minimum
+    requirements whilst giving authors the ability to tweak the tests as necessary
+    for their specific adapter.
+    """
+    @property
+    @abstractmethod
+    def tadapter(self):
+        """Adapter to test
+        """
+        pass
+
+    def test_initialize(self):
+        tadapter = self.tadapter
+        assert not tadapter.initialized
+        tadapter._initialize_model()
+        assert tadapter.initialized
+
+    def test_shutdown(self, test_adapter):
+        """Test the adapter can be shutdown
+
+        Extra tests can be adapted depending on what the adapter should actually
+        do on shutdown.
+        """
+        test_adapter.shutdown()
+
+    def test_initialise_model_input(self, test_adapter, test_config_paraset):
+        """Test that initalising model input does as intended
+        """
+        # @swillner I am not sure how we want this to work. I still don't understand
```

Comment:
I think that should be up to the model/adapter. Maybe we should also implement a way for the adapter to tell, e.g. for an ensemble run, if it would like to change the model input or the run parameters faster in the ensembke loop.

Reply:
> Maybe we should also implement a way for the adapter to tell

That sounds good, thoughts on how to do it that don't require passing a parameter set to these calls? A flag (which by default signals that it's a run parameter but the user can switch to model input for speed if they want?)

Reply:
I think we need both, one for the adapter indicating the preference of order of these two functions (in an ensemble loop), and something indicating for the paramters. The latter we could specify as a standard as that should be model independent, isn’t it?

Reply:
> I think we need both, one for the adapter indicating the preference of order of these two functions (in an ensemble loop)

I think it will always be `set_model_inputs`, then `set_run_parameters` (timeseries are longer and involve more data than configuration flags so you want to do timeseries setting less often). 

> The latter we could specify as a standard as that should be model independent, isn’t it?

Yep!

So perhaps we have two standards to add. 1) That models know that `set_model_inputs` should be used for 'scenario/driver setting' and so would be called less often than `set_run_parameters` which would be used for 'config setting'. 2) Parameters have a flag indicating whether they should be set when `set_model_inputs` is called or whether they should be set when `set_run_parameters` is called and adapters have to know how to handle this?

Reply:
cool. If we set a standard we don’t need a flag, it’s up to the user/adapter dev to adhere to them

Reply:
Ok I'll leave that to you to as I still don't really get how to implement it